### ISSUE TITLE
list() in foreach(): wrong type inferred

### ISSUE THREAD

**Issue by donquixote** [kalessil/phpinspectionsea#718]
Example code:

```php
  /**
   * @param int[][] $items
   */
  function foo(array $items) {
    $result = [];
    foreach ($items as list($k0, $k1, $k2)) {
      $result[$k0] = 5;
      $result[$k1] = 5;
      $result[$k2] = 5;
    }
  }
```

Strangely, in the last of the 3 assignments, the `$k2` produces an inspection warning:

> Array and string offset validity
> Resolved index type ([array]) is incompatible with possible [string, int]. Probably just proper type hinting needed.

![phpstorm-eaextended-list-in-foreach](https://user-images.githubusercontent.com/150032/34311657-1099d258-e75f-11e7-96cc-b2f4c3834066.png)

So it seems as if the plugin thinks that $k0 and $k2 are integer, but $k3 is not.
I don't know if this is EA Extended plugin or if it is PhpStorm core.

**Comment by donquixote** [comment]
> I don't know if this is EA Extended plugin or if it is PhpStorm core.

I mean the inspection is from EA Extended. But the type sniffing might be core behavior?

**Comment by donquixote** [comment]
I can change it like so:

```php
  /**
   * @param int[][] $items
   */
  function foo(array $items) {
    $result = [];
    foreach ($items as $item) {
      list($k0, $k1, $k2) = $item;
      $result[$k0] = 5;
      $result[$k1] = 5;
      $result[$k2] = 5;
    }
  }
```

This eliminates the problem on `$k2`. But then I get

> 'list(...) = ' usage possible
> foreach (... as list(...)) could be used instead.

**Comment by kalessil** [comment]
Confirmed, types are not resolved correctly:
- k0: int|string
- k1: <no type information>
- k2: int[]|mixed

Please report this case to JetBrains.

I also checked if I can introduce a workaround and to not mess the inspection code out, but this will not work.

**Comment by donquixote** [comment]
https://youtrack.jetbrains.com/issue/WI-39365

**Comment by kalessil** [comment]
Thank you @donquixote
